test/integration/master/kube_apiserver_test.go

```diff
@@ -85,6 +85,83 @@ func TestRun(t *testing.T) {
 	}
 }
 
+type delayedCheck struct {
+	startPassingAfter time.Time
+}
+
+func (h delayedCheck) Name() string {
+	return "fail-for-ten-seconds"
+}
+
+func (h delayedCheck) Check(req *http.Request) error {
+	if time.Now().After(h.startPassingAfter) {
+		return nil
+	}
+	return fmt.Errorf("failing until after %v", h.startPassingAfter.Unix())
+}
+
+func TestDelayedHealthzAndReadyz(t *testing.T) {
+	hc := delayedCheck{time.Now().Add(10 * time.Second)}
+	instanceOptions := &kubeapiservertesting.TestServerInstanceOptions{
+		InjectedHealthzChecker: hc,
+	}
+	server := kubeapiservertesting.StartTestServerOrDie(t, instanceOptions, []string{"--post-start-hook-grace-period", "5s"}, framework.SharedEtcd())
+	defer server.TearDownFn()
+
+	client, err := kubernetes.NewForConfig(server.ClientConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	readyzRes := client.CoreV1().RESTClient().Get().AbsPath("/readyz").Do()
+	readyzStatus := 0
+	readyzRes.StatusCode(&readyzStatus)
+	if readyzStatus == 200 {
+		t.Fatalf("Readyz should start unhealthy")
+	}
+	errCh := make(chan error)
+	err = wait.Poll(100*time.Millisecond, 10*time.Second, func() (bool, error) {
```

Comment:
hardcoded times are suspicious

Reply:
we need to wait longer than the grace period. i will get rid of the other polling instances though.